README.md
# Animated Menu Button

## Introduction

This is provide the animated button that can help ui further your UI jouney


### Demo
https://drive.google.com/file/d/1YVmRYFRv36nfILjMHtaE8mJyDDRiiDzv/view?usp=sharing

## Example
```dart
# File: example.dart

import 'package:flutter/material.dart';
import 'package:rotate_pills_animation/components/animated_,menu_button.dart';

class ExamplePage extends StatelessWidget {
  const ExamplePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: [
          const Expanded(
            child: Center(
              child: Text("Hello world"),
            ),
          ),
          SizedBox(
            height: 100,
            child: Row(
              children: [
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
                Expanded(
                    child: Center(
                  child: AnimatedMenuButton(
                    radius: 50,
                    homeWidget: item(
                      color: Colors.black,
                      child: const Icon(
                        Icons.add,
                        color: Colors.white,
                      ),
                    ),
                    actionWidgets: [
                      InkWell(
                        onTap: () {
                          print("add");
                        },
                        child: item(
                          color: Colors.green,
                          child: const Icon(Icons.face),
                        ),
                      ),
                      InkWell(
                        onTap: () {
                          print("image");
                        },
                        child: item(
                          color: Colors.blue,
                          child: const Icon(Icons.image),
                        ),
                      ),
                      InkWell(
                        onTap: () {
                          print("cached_rounded");
                        },
                        child: item(
                          color: Colors.red,
                          child: const Icon(Icons.cached_rounded),
                        ),
                      )
                    ],
                    animatedDuration: const Duration(milliseconds: 1000),
                  ),
                )),
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
              ],
            ),
          )
        ],
      ),
    );
  }

  Widget item({required Widget child, required Color color}) {
    return Container(
      padding: const EdgeInsets.all(5),
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        color: color,
      ),
      child: child,
    );
  }
}



CHANGELOG.md
## 1.0.0
add package
## 1.0.1
format code


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:rotate_pills_animation/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget( MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "rotate_pills_animation");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "rotate_pills_animation");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"rotate_pills_animation", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/components/animated_,menu_button.dart
import 'dart:math';
import 'package:flutter/material.dart';

class AnimatedMenuButton extends StatefulWidget {
  final Widget homeWidget;
  //only accept 3 widgets
  final List<Widget> actionWidgets;
  final double radius;
  final Duration animatedDuration;
  const AnimatedMenuButton({super.key, required this.homeWidget, required this.actionWidgets, this.radius = 100, required this.animatedDuration});

  @override
  State<AnimatedMenuButton> createState() => _AnimatedMenuButtonState();
}

class _AnimatedMenuButtonState extends State<AnimatedMenuButton> with TickerProviderStateMixin{
  late AnimationController _item1Controller;
  late Animation<double> _item1animation;
  late AnimationController _item2Controller;
  late Animation<double> _item2Animation;
  late AnimationController _item3Controller;
  late Animation<double> _item3Animation;

  bool isOpen = false;
  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _item1Controller = AnimationController(vsync: this, duration: widget.animatedDuration);
    _item1animation = Tween<double>(begin: -pi, end: -pi * 0.2).animate(_item1Controller);


    _item2Controller = AnimationController(vsync: this, duration: widget.animatedDuration);
    _item2Animation = Tween<double>(begin: -pi, end: -pi/2).animate(_item2Controller);


    _item3Controller = AnimationController(vsync: this, duration: widget.animatedDuration);
    _item3Animation = Tween<double>(begin: -pi, end: -pi * 0.8).animate(_item3Controller);
    _item1Controller.addStatusListener((status) {
      if(status == AnimationStatus.completed){

      }
    });
  }

  @override
  void dispose() {
    _item1Controller.dispose();
    _item2Controller.dispose();
    _item3Controller.dispose();
    super.dispose();
  }

  Widget item(bool visible, Widget child){
    return Visibility(
        visible: visible,
        child: child);
  }
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Stack(
        children: [
          InkWell(
            onTap: (){
              if(!isOpen){
                isOpen = !isOpen;
                _item1Controller.forward();
                _item2Controller.forward();
                _item3Controller.forward();
              } else {
                _item1Controller.reverse();
                _item2Controller.reverse();
                _item3Controller.reverse();
                Future.delayed(widget.animatedDuration, (){setState(() {
                  isOpen = !isOpen;
                });},);
              }



            },
            child: item(true, widget.homeWidget),
          ),
          AnimatedBuilder(
              animation: _item3Controller,
              builder: (context, child){
                var offset = getOffset(const Offset(0, 0),  _item3Animation.value);
                return Transform(
                  transform: Matrix4.identity()
                    ..translate(offset.dx, offset.dy
                    ),
                  child: item(isOpen, widget.actionWidgets[0]),
                );
              }
          ),
          AnimatedBuilder(
              animation: _item2Controller,
              builder: (context, child){
                var offset = getOffset(const Offset(0, 0), _item2Animation.value);
                return Transform(
                  transform: Matrix4.identity()
                    ..translate(offset.dx, offset.dy
                    ),
                  child: item(isOpen, widget.actionWidgets[1]),
                );
              }
          ),
          AnimatedBuilder(
              animation: _item1Controller,
              builder: (context, child){
                var offset = getOffset(const Offset(0, 0),_item1animation.value);
                return Transform(
                  transform: Matrix4.identity()
                    ..translate(offset.dx, offset.dy
                    ),
                  child: item(isOpen, widget.actionWidgets[2]),
                );
              }
          )
        ],
      ),
    );
  }

  Offset getOffset(Offset center, double angle){
    double x = center.dx + widget.radius * cos(angle);
    double y = center.dy + widget.radius * sin(angle);

    return Offset(x, y);
  }
}


lib/example.dart
import 'package:flutter/material.dart';

import 'components/animated_,menu_button.dart';

class ExamplePage extends StatelessWidget {
  const ExamplePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: [
          const Expanded(
            child: Center(
              child: Text("Hello world"),
            ),
          ),
          SizedBox(
            height: 100,
            child: Row(
              children: [
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
                Expanded(
                    child: Center(
                  child: AnimatedMenuButton(
                    radius: 50,
                    homeWidget: item(
                      color: Colors.black,
                      child: const Icon(
                        Icons.add,
                        color: Colors.white,
                      ),
                    ),
                    actionWidgets: [
                      InkWell(
                        onTap: () {
                        },
                        child: item(
                          color: Colors.green,
                          child: const Icon(Icons.face),
                        ),
                      ),
                      InkWell(
                        onTap: () {
                        },
                        child: item(
                          color: Colors.blue,
                          child: const Icon(Icons.image),
                        ),
                      ),
                      InkWell(
                        onTap: () {
                        },
                        child: item(
                          color: Colors.red,
                          child: const Icon(Icons.cached_rounded),
                        ),
                      )
                    ],
                    animatedDuration: const Duration(milliseconds: 1000),
                  ),
                )),
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
                const SizedBox(
                  width: 60,
                  child: Icon(Icons.ac_unit),
                ),
              ],
            ),
          )
        ],
      ),
    );
  }

  Widget item({required Widget child, required Color color}) {
    return Container(
      padding: const EdgeInsets.all(5),
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        color: color,
      ),
      child: child,
    );
  }
}
